File: daily-coding-problem/dcp-py/1_20/problem13_longest_substring.py
"""
This problem was asked by Amazon.

Given an integer k and a string s, find the length of the longest substring that contains at most k distinct characters.

For example, given s = "abcba" and k = 2, the longest substring with k distinct characters is "bcb".
"""


def longest_substring(k, s):
    def find(rem, count, longest, acc):
        if not rem:
            return longest
        elif rem[0] in acc:
            return find(rem[1:], count, longest, acc + rem[0])
        elif rem[0] not in acc and count < k:
            return find(rem[1:], count + 1, longest, acc + rem[0])
        else:
            return find(acc[-1] + rem, 0, max_len(longest, acc), "")  # reset

    return find(s, 0, "", "") if k > 1 else s[0]


def max_len(s1, s2):
    return s1 if len(s1) > len(s2) else s2


assert longest_substring(2, 'abcba') == 'bcb'
assert longest_substring(3, 'abbbbcda') == 'abbbbc'
assert longest_substring(1, 'abcbbbaaaaaaaaaabbcbbadd') == 'a'

File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem13_LongestSubstring.scala
package dcp.problems._1_20

/**
 * This problem was asked by Amazon.
 *
 * Given an integer k and a string s, find the length of the longest substring that contains at most k distinct
 * characters.
 *
 * For example, given s = "abcba" and k = 2, the longest substring with k distinct characters is "bcb".
 */

object Problem13_LongestSubstring {

  def solution(k: Int, s: String): String = {
    @scala.annotation.tailrec
    def findLongest(rem: String, acc: String = "", longest: String = ""): String = {
      if (rem.isEmpty) longest
      else {
        val count = acc.toSet.size
        val c = rem.charAt(0)
        if (acc.contains(c)) findLongest(rem.substring(1), acc + c, longest)
        else if (!acc.contains(c) && count < k) findLongest(rem.substring(1), acc + c, longest)
        else findLongest(acc.substring(acc.length - 1) + rem, "", maxLen(acc, longest))
      }
    }

    if (k > 1) findLongest(s) else s(0).toString
  }

  def maxLen(s1: String, s2: String): String = if (s1.length > s2.length) s1 else s2
}
